eth/filters/api.go

```diff
@@ -269,19 +330,94 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
 			case logs := <-matchedLogs:
 				for _, log := range logs {
 					log := log
-					notifier.Notify(rpcSub.ID, &log)
+					notify.Notify(rpcSub.ID, &log)
 				}
 			case <-rpcSub.Err(): // client send an unsubscribe request
 				logsSub.Unsubscribe()
 				return
-			case <-notifier.Closed(): // connection dropped
+			case <-notify.Closed(): // connection dropped
 				logsSub.Unsubscribe()
 				return
 			}
 		}
 	}()
+	return nil
+}
 
-	return rpcSub, nil
+// histLogs only retrieves the logs older than current header
+func (api *FilterAPI) histLogs(ctx context.Context, notifier notifier, rpcSub *rpc.Subscription, n int64, crit FilterCriteria) (int64, error) {
+	// the ctx will be canceled after this callback(filter.Logs) is called
+	// and we are run in a background goroutine, use a new context instead
+	var (
+		cctx     = context.Background()
+		rmLogsCh = make(chan []*types.Log)
+	)
+	for {
+		// get the latest block header
+		header := api.sys.backend.CurrentHeader()
+		if header == nil {
+			return 0, errors.New("header is null")
+		}
+		head := header.Number.Int64()
+		if n >= head {
+			return head, nil
+		}
+
+		// do historical sync from n to head
+		f := api.sys.NewRangeFilter(n, head, crit.Addresses, crit.Topics)
+		logChan, errChan := f.rangeLogsAsync(cctx)
+
+		// subscribe rmLogs
+		query := ethereum.FilterQuery{FromBlock: big.NewInt(n), ToBlock: big.NewInt(head), Addresses: crit.Addresses, Topics: crit.Topics}
```

Comment:
Took me a minute to get this, but smart trick I love it!

Reply:
Reality check: It's more complicated. 

- Reorg happens in a future block that `rangeLogsAsync` hasn't processed yet. In that case we don't really care about the event since when `rangeLogsAsync` reaches that point it will go on the right chain by itself.
- We can keep track of how far we are in delivery by checking the block number of incoming logs.
- However it's harder to say where is the point of reorg, because removed logs are sent in the reverse order (from more recent block backwards).
- What we can do is compare as they come in to see if they pass the blocks for which we've already delivered logs.
- At this point we have to stop `rangeAsyncLogs` immediately. Or hmm there's a question what should be done if it is mid-block (with some logs remaining in that block).
- Then the new subscription which sends removed logs will also send the replacement logs of the new chain. So we can just tune into that.